eta/core/serial.py

```diff
@@ -670,32 +705,325 @@ def from_dict(cls, d):
+    directory on disk. Any accessing of an element in the list causes a READ
+    from disk and updating any elements in the list REQUIRES setting the item
+    explicitly (and causes a WRITE to disk).
+
+    This class cannot be instantiated directly.
+
+    A list of uuids is managed by the DirectoryContainer in _ELE_ATTR to keep
+    order and find elements in the directory. This list a part of
+    serialization. The data directory location is also serialized in
+    _ELE_DIR_ATTR.
+
+    USE WITH CARE.
+
+    Usage:
+        element = cls._ELE_CLS()
+        c = DirectoryContainerSubclass.from_json("...")
+
+        # All getters cause an IMMEDIATE READ from disk for each element
+        element = c[0]
+        in_memory_list = [e for e in c] # NOT recommended
+
+        # The setters cause an IMMEDIATE WRITE to disk
+        c[0] = element
+        c.append(element)
+        c.add(element)
+    '''
+
+    #
+    # The name of the attribute that will store the path to the json data files
+    #
+    # Subclasses MUST set this field
+    #
+    _ELE_DIR_ATTR = None
+
+    def __init__(self, **kwargs):
+        '''Creates a DirectoryContainer instance.
+
+        NOTE: The <elements> uuid list is generated dynamically from the path if
+        it is not provided.
+
+        Args:
+            <elements_dir>: the directory path to store the data. This argument
+                is required. The appropriate name of this keyword argument is
+                determined by the `_ELE_DIR_ATTR` member of the
+                DirectoryContainer subclass
+            <elements>: an optional list of element uuids. globbing is done when
+                this is not provided. The appropriate name of this keyword
+                argument is determined by the `_ELE_ATTR` member of the
+                DirectoryContainer subclass
+
+        Raises:
+            ContainerError: if there was an error while creating the container
+        '''
+        self._validate()
+
+        if self._ELE_DIR_ATTR not in kwargs:
+            raise ContainerError(
+                "Missing keyword argument %s is required"
+                % (self._ELE_DIR_ATTR))
+
+        setattr(self, self._ELE_DIR_ATTR, kwargs[self._ELE_DIR_ATTR])
+
+        setattr(
+            self,
+            self._ELE_ATTR,
+            kwargs.get(
+                self._ELE_ATTR,
+                [
+                    os.path.splitext(os.path.basename(i))[0]
+                    for i in etau.multiglob(".json",
+                                            root=self._ele_dir+"/**/*")
+                ]
+            )
+        )
+
+    def __iter__(self):
+        return (self._load_ele(path) for path in self._paths)
+
+    def __getitem__(self, idx):
+        return self._load_ele(self._ele_path(idx))
+
+    def __setitem__(self, idx, ele):
+        ele.write_json(self._ele_path(idx))
+
+    def __delitem__(self, idx):
+        etau.delete_file(self._ele_path(idx))
```

Comment:
Do we want delete to actually delete the file?  I could see cases where we would not want to include the item in the list anymore, but not actually want to delete the underlying file.